### DISCUSSION TITLE
导出模型inference后，paddleocr命令行执行结果不一致

### DISCUSSION THREAD

**Discussion by mxdg** [PaddlePaddle/PaddleOCR#13973] [Q&A]
导出模型inference后，同样的模型，同样的图片，执行结果不一致（有几个字母不一样），是有参数不对吗？

这里执行预测结果正确：
python tools/infer/predict_rec.py --image_dir="/root/autodl-tmp/jg3.png" --rec_model_dir="/root/autodl-tmp/mydata/v4_hgnet_1/" --rec_char_dict_path="/root/autodl-tmp/mydata/kaz_dict.txt" --use_gpu=False
![image](https://github.com/user-attachments/assets/1e2d573e-d86e-4b0b-a29f-cfa56719989a)

这里执行结果有错误：
paddleocr --image_dir /root/autodl-tmp/jg3.png --use_angle_cls true --use_gpu false --rec_model_dir="/root/autodl-tmp/mydata/v4_hgnet_1/" --rec_char_dict_path="/root/autodl-tmp/mydata/kaz_dict.txt"
![image](https://github.com/user-attachments/assets/0489e668-589f-459f-8b45-35c991454037)

python 3.10

Name: paddleocr
Version: 2.8.1

**Comment by karanpate1** [comment]
> After exporting the model inference, the execution results are inconsistent (a few letters are different) for the same model and the same image. Is there something wrong with the parameters?
> 
> The prediction result is correct here: python tools/infer/predict_rec.py --image_dir="/root/autodl-tmp/jg3.png" --rec_model_dir="/root/autodl-tmp/mydata/v4_hgnet_1/" --rec_char_dict_path="/root/autodl-tmp/mydata/kaz_dict.txt" --use_gpu=False ![image](https://private-user-images.githubusercontent.com/30197124/375399687-1e2d573e-d86e-4b0b-a29f-cfa56719989a.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTg3OTQwMzEsIm5iZiI6MTc1ODc5MzczMSwicGF0aCI6Ii8zMDE5NzEyNC8zNzUzOTk2ODctMWUyZDU3M2UtZDg2ZS00YjBiLWEyOWYtY2ZhNTY3MTk5ODlhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTI1VDA5NDg1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNmYzU4Mjc5NGI2ZjcyYjhiZDBhYzQ5NDA4NWY5NmE0N2RhNzQ4ZDQ4ZmI3MzRmOGYyOWQ0N2RlYTZkMzQ3YzMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.cI7UWkos2rk60cnRWt9VStG2jYPCFgzZpv92IST3OIw)
> 
> The execution result here is wrong: paddleocr --image_dir /root/autodl-tmp/jg3.png --use_angle_cls true --use_gpu false --rec_model_dir="/root/autodl-tmp/mydata/v4_hgnet_1/" --rec_char_dict_path="/root/autodl-tmp/mydata/kaz_dict.txt" ![image](https://private-user-images.githubusercontent.com/30197124/375399720-0489e668-589f-459f-8b45-35c991454037.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTg3OTQwMzEsIm5iZiI6MTc1ODc5MzczMSwicGF0aCI6Ii8zMDE5NzEyNC8zNzUzOTk3MjAtMDQ4OWU2NjgtNTg5Zi00NTlmLThiNDUtMzVjOTkxNDU0MDM3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTI1VDA5NDg1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI1OTk2MzY2ZDZlYWZiYjQ5MmQ0M2RiNmMzOGYyYWY2Njc1MjJkMDczMGY2N2QyMzNhM2NiOTg5MWFlYjJiN2EmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.y1N4aOFRBvDbTXaQZoXBCrB0o_iXlwMyrIIMAaLBuXs)
> 
> python 3.10
> 
> Name: paddleocr Version: 2.8.1


###  Issue: High Validation Accuracy but Garbage Inference in PP-OCRv5 (Solved)

I fine-tuned **PP-OCRv5** on a custom dataset (English alphabets, numerals, and some symbols).

* Training accuracy: **98%**
* Validation accuracy: **99%**

But when I ran inference (even on images from training/validation sets), the predictions were totally wrong.

I double-checked everything:

* Followed official finetuning + export steps
* Tested with both `tools/infer_rec.py` and exported inference model
* Same problem: **terrible results**

At this point, I was stuck because evaluation showed **99% accuracy**, but inference was useless.

---

###  Root Cause

Turns out the problem was due to **mismatch in image preprocessing between training and inference**.

In my training config `PP-OCRv5_mobile_rec.yml`:

```yaml
sampler:
  name: MultiScaleSampler
  scales: [[320, 32], [320, 48], [320, 64]]
```

So during training, images were dynamically resized to **3 different shapes**:

* 3×32×320
* 3×48×320
* 3×64×320

But in my exported `inference.yml`, the preprocessing looked like this:

```yaml
PreProcess:
  transform_ops:
  - DecodeImage:
      channel_first: false
      img_mode: BGR
  - MultiLabelEncode:
      gtc_encode: NRTRLabelEncode
  - RecResizeImg:
      image_shape:
      - 3
      - 48
      - 320
```

Notice the mismatch?

* Training: random **multi-scale** (32, 48, 64 heights)
* Inference: fixed to **48 height only**

That mismatch caused the network to completely fail at inference.

---

### ✅ Fix

I manually adjusted the `inference.yml` to try different scales from training.

* With **3×48×320**, results were trash.
* With **3×64×320**, also trash.
* With **3×32×320**, predictions became **perfect**, matching my training/validation accuracy.

So the fix was:
 **Set `RecResizeImg.image_shape` in `inference.yml` to the same shape(s) used in training.**

Final fix:

```yaml
PreProcess:
  transform_ops:
  - DecodeImage:
      channel_first: false
      img_mode: BGR
  - MultiLabelEncode:
      gtc_encode: NRTRLabelEncode
  - RecResizeImg:
      image_shape:
      - 3
      - 32
      - 320
```
---

###  Takeaway

If you’re fine-tuning PP-OCR models with `MultiScaleSampler`, make sure your **inference preprocessing matches your training scales**. Otherwise, you’ll see high validation accuracy but garbage results at inference.

---

This should save people from wasting days like I did. 🙃

---

Want me to polish this into a proper **GitHub issue comment** format (with Markdown headings + code blocks, short and punchy)? Or do you want it more like a **blog-style post** with more storytelling?
